torch/nn/modules/transformer.py

```diff
@@ -0,0 +1,519 @@
+            Note: Due to the multi-head attention architecture in the transformer model, 
+                  the output sequence length of a transformer is same as the input sequence
+                  (i.e. target) length of the decode. 
+
+        Examples:
+            >>> output = transformer_model(src, tgt, src_mask=src_mask, tgt_mask=tgt_mask)
+        """
+
+        memory = self.encode(src, src_mask)
+        output = self.decode(tgt, memory, tgt_mask, memory_mask)
+
+        if self.generator:
+            output = self.generator(output)
+
+        return output
+
+    def encode(self, src, src_mask=None):
+        return self.encoder(src, src_mask)
+
+    def decode(self, tgt, memory, tgt_mask=None, memory_mask=None):
+        return self.decoder(tgt, memory, tgt_mask, memory_mask)
+
+    def generate_square_subsequent_mask(self, sz):
+        r"""Generate a square mask for the sequence. The masked positions are filled with float('-inf').
+            Unmasked positions are filled with float(0.0)
+        """
+        attn_shape = (sz, sz)
+        mask = np.triu(np.ones(attn_shape), k=1).astype('uint8')
+        mask = torch.from_numpy(mask) == 0
+        mask = mask.float().masked_fill(mask == 0, float('-inf')).masked_fill(mask == 1, float(0.0))
+        return mask    
+
+    def _reset_parameters(self):
+        r"""Initiate parameters in the transformer model."""
+
+        for p in self.parameters():
+            if p.dim() > 1:
+                xavier_uniform_(p)
+
+
+class TransformerEncoder(Module):
+    r"""TransformerEncoder is a stack of N encoder layers
+
+    Args:
+        encoder_layer: an instance of the TransformerEncoderLayer() class (required). 
+        num_layers: the number of sub-encoder-layers in the encoder (required).
+        src_embed: the embedding model for the source sequence (optional).
+        pos_encoder: the positional encoding model for the source sequence (optional).
+        norm: the layer normalization component (optional).
+
+    Examples::
+        >>> transformer_encoder = nn.TransformerEncoder(encoder_layer, num_layers)
+    """
+
+    def __init__(self, encoder_layer, num_layers, src_embed=None, pos_encoder=None, norm=None):
+        super(TransformerEncoder, self).__init__()
+        self.layers = _get_clones(encoder_layer, num_layers)
+        self.num_layers = num_layers
+        self.src_embed = src_embed
+        self.pos_encoder = pos_encoder
+        self.norm = norm
+
+    def forward(self, src, mask=None):
+        r"""Pass the input through the endocder layers in turn.
+
+        Args:
+            src: the sequnce to the encoder (required).
+            src_mask: the mask for the src sequence (optional).
+
+        Shape:
+            see the docs in Transformer class.
+        """
+        output = src
+        if self.src_embed:
+            output = self.src_embed(output)
+
+        if self.pos_encoder:
+            output = self.pos_encoder(output)
```

Comment:
You can assume that the user is passing the result of something akin to "self.src_embed(self.pos_encoder(src))" directly as the input (src). This is not something I'd add statically. We usually require the user to do the embedding herself and then pass in the result vector to our models.